File: sdk-go/vconn/capstate.go
package vconn

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// CapabilityState is the client's local record of the most recent successful
// service:"certificate" exchange (docs/CERTIFICATE_WEB_OF_TRUST.md §3.2b).
// The Challenger certificate it carries is a native, independently
// verifiable artifact — not session-bound like the auth handshake — so
// persisting it here (unlike the in-memory grant it's read into) means a
// process restart resumes with the same capabilities instead of losing them
// until an admin re-triggers delivery.
type CapabilityState struct {
	CertID       string   `json:"certId"`
	Certificate  string   `json:"certificate"`
	Capabilities []string `json:"capabilities"`
}

// loadCapabilityState reads persisted state from path, returning (nil, nil)
// if path is empty or the file doesn't exist yet — both are the normal
// "nothing granted so far" case, not an error.
func loadCapabilityState(path string) (*CapabilityState, error) {
	if path == "" {
		return nil, nil
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("vconn: reading capability state: %w", err)
	}
	var state CapabilityState
	if err := json.Unmarshal(data, &state); err != nil {
		return nil, fmt.Errorf("vconn: parsing capability state: %w", err)
	}
	return &state, nil
}

// saveCapabilityState writes state to path (mode 0600, matching the identity
// secret's own permissions), creating its parent directory if needed. A
// no-op when path is empty.
func saveCapabilityState(path string, state CapabilityState) error {
	if path == "" {
		return nil
	}
	if dir := filepath.Dir(path); dir != "." && dir != "" {
		if err := os.MkdirAll(dir, 0o700); err != nil {
			return fmt.Errorf("vconn: creating capability state directory: %w", err)
		}
	}
	data, err := json.Marshal(state)
	if err != nil {
		return fmt.Errorf("vconn: encoding capability state: %w", err)
	}
	return os.WriteFile(path, data, 0o600)
}

File: packages/sdk/src/cap-state.ts
/**
 * Local persistence of successful certificate exchanges.
 *
 * Worth persisting because the Challenger certificate is a native, independently
 * verifiable artefact rather than something session-bound: keeping it means a
 * process restart resumes with the same certificates, instead of holding nothing
 * until an admin happens to re-trigger delivery.
 */

import fs from "node:fs";
import path from "node:path";
import { expandHome } from "./identity.js";

export interface CapabilityStateCertificate {
  certId: string;
  certificate: string;
  capabilities: string[];
  /**
   * The status the control plane last reported for this certificate, and when.
   *
   * Absent means "never checked", which is treated as maximally stale rather than as active:
   * assuming active is what made revocation invisible before status checking existed at all.
   */
  lastStatus?: "active" | "revoked" | "superseded" | "expired";
  /** Ms since epoch of the last verified status response. */
  lastCheckedAt?: number;
  issuedAt?: number;
  expiresAt?: number | null;
  resourceLimits?: unknown;
  scope?: unknown;
}

export interface CapabilityState {
  certificates: CapabilityStateCertificate[];
}

/**
 * Read persisted state.
 *
 * Returns `null` for both "no path configured" and "file does not exist yet" —
 * both are the ordinary "nothing granted so far" case, not an error. A malformed
 * file *is* an error: silently discarding it would look identical to a fresh
 * install and quietly drop authority the Actor legitimately holds.
 */
export function loadCapabilityState(statePath?: string): CapabilityState | null {
  if (!statePath) return null;
  const resolved = expandHome(statePath);
  if (!fs.existsSync(resolved)) return null;

  const raw = fs.readFileSync(resolved, "utf-8");
  const parsed = JSON.parse(raw) as CapabilityState;
  if (!parsed || !Array.isArray(parsed.certificates)) {
    throw new Error(`Malformed capability state at ${resolved}`);
  }
  for (const cert of parsed.certificates) {
    if (!cert || typeof cert.certId !== "string" || !Array.isArray(cert.capabilities)) {
      throw new Error(`Malformed capability state at ${resolved}`);
    }
  }
  return parsed;
}

/** Write state with mode 0600, matching the identity secret's permissions. No-op without a path. */
export function saveCapabilityState(statePath: string | undefined, state: CapabilityState): void {
  if (!statePath) return;
  const resolved = expandHome(statePath);
  const dir = path.dirname(resolved);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
  fs.writeFileSync(resolved, JSON.stringify(state), { encoding: "utf-8", mode: 0o600 });
}
